#---------------------------------------------------------------------
# Script Name:   PHGFunctions.R
# Description:   Functions to support PHG
# Author:        Brandon Monier & Ed buckler
# Created:       2018-11-26 at 11:14:36
# Last Modified: 2018-12-03 at 17:58:46
#--------------------------------------------------------------------

#--------------------------------------------------------------------
# Detailed Purpose:
#    The main purpose of this Rscript to support the autogenerated wrappers
#   for TASSEL plugins in TasselPluginWrappers
#--------------------------------------------------------------------

convertHaplotypesDataVectorsToDataFrame <- function(graph, refRanges, includeSequence, includeVariants) {
  hapDataVectors <- J("net.maizegenetics.pangenome.pipelineTests.GenerateRForPHG")$graphToHapsInRefRangeVectors(graph, refRanges, includeSequence, includeVariants)
  cnames <- unlist(strsplit("hapids,refRangeIds,methodIds,taxa,sequence,variantInfo", ","))
  colNum <- length(cnames)
  colNum
  aDF <- data.frame(hapDataVectors$dataVector$get(0L))
  for(i in 2:(colNum)) {
    aDF[[i]] <- hapDataVectors$dataVector$get(i-1L)
  }
  colnames(aDF) <- cnames
  aDF
}

convertRefRangeVectorsToDataFrame <- function(graph, refRanges) {
  refRangeVectors <- J("net.maizegenetics.pangenome.pipelineTests.GenerateRForPHG")$graphToRefRangeVectors(graph, refRanges)
  cnames <- unlist(strsplit("refRangeIds,chromosomes,startPos,endPos,refLineName,numberOfNodes", ","))
  colNum <- length(cnames)
  colNum
  aDF <- data.frame(refRangeVectors$dataVector$get(0L))
  for(i in 2:(colNum)) {
    aDF[[i]] <- refRangeVectors$dataVector$get(i-1L)
  }
  colnames(aDF) <- cnames
  aDF
}