#---------------------------------------------------------------------
# Script Name:   PHGPluginWrappers.R
# Description:   Various tests with rJava
# Author:        Ed Buckler & Lynn Johnson
# Created:       2018-11-26 at 11:14:36
# Last Modified: 2018-12-03 at 17:58:46
#--------------------------------------------------------------------

#--------------------------------------------------------------------
# Detailed Purpose:
#    The main purpose of this Rscript to host to autogenerated wrappers
#   for TASSEL plugins
#--------------------------------------------------------------------


#Test PHG usage in R
## Set WD
setwd("~/Code/rtassel")
path_exp_PHG <- paste0(getwd(),"/inst/java/PHG.jar")

rJava::.jaddClassPath(path_exp_PHG)
print(.jclassPath())



haplotypeGraphBuilderPlugin <- function(configFile, myMethods
) {
  plugin <- new(J("net.maizegenetics.pangenome.api.HaplotypeGraphBuilderPlugin"), .jnull(), FALSE)
  plugin$setParameter("configFile",toString(configFile))
  plugin$setParameter("myMethods",toString(myMethods))
  plugin$setParameter("myIncludeSequences",toString(FALSE))
  plugin$build()
}

haplotypeGraphStreamBuilderPlugin <- function(configFile
) {
  plugin <- new(J("net.maizegenetics.pangenome.api.HaplotypeGraphStreamBuilderPlugin"), .jnull(), FALSE)
  plugin$runPlugin(configFile)
}

configFilePath <- paste0(getwd(),"/data/configSQLiteR.txt")
method <- "B73ref"
plugin <- new(J("net.maizegenetics.pangenome.api.HaplotypeGraphBuilderPlugin"), .jnull(), FALSE)
#plugin$setParameter("configFile",toString(configFile))
#plugin$setParameter("myMethods",toString(myMethods))
# plugin$setParameter("myIncludeSequences",toString(FALSE))
plugin$configFile(configFilePath)
plugin$methods(method)
jfalse <- .jnew("java/lang/Boolean",FALSE)
jfalse
plugin$includeSequences(jfalse)
hg <- plugin$build()
hg$chromosomes()
rrList <- hg$referenceRangeList()
rrList$get(0L)


hn <- hg$nodes

haplotypeGraphBuilderPlugin(configFilePath,method)
